javalib/src/main/scala/java/util/ArrayDeque.scala

```diff
@@ -0,0 +1,241 @@
+
+  def removeFirst(): E = {
+    if (inner.isEmpty)
+      throw new NoSuchElementException()
+    else
+      pollFirst()
+  }
+
+  def removeLast(): E = {
+    if (inner.isEmpty)
+      throw new NoSuchElementException()
+    else
+      pollLast()
+  }
+
+  def pollFirst(): E = {
+    if (inner.isEmpty) null.asInstanceOf[E]
+    else {
+      val res = inner.remove(0)
+      status += 1
+      res
+    }
+  }
+
+  def pollLast(): E = {
+    if (inner.isEmpty) null.asInstanceOf[E]
+    else inner.remove(inner.size - 1)
+  }
+
+  def getFirst(): E = {
+    if (inner.isEmpty)
+      throw new NoSuchElementException()
+    else
+      peekFirst()
+  }
+
+  def getLast(): E = {
+    if (inner.isEmpty)
+      throw new NoSuchElementException()
+    else
+      peekLast()
+  }
+
+  def peekFirst(): E = {
+    if (inner.isEmpty) null.asInstanceOf[E]
+    else inner.get(0)
+  }
+
+  def peekLast(): E = {
+    if (inner.isEmpty) null.asInstanceOf[E]
+    else inner.get(inner.size - 1)
+  }
+
+  def removeFirstOccurrence(o: Any): Boolean = {
+    val index = inner.indexOf(o)
+    if (index >= 0) {
+      inner.remove(index)
+      status += 1
+      true
+    } else
+      false
+  }
+
+  def removeLastOccurrence(o: Any): Boolean = {
+    val index = inner.lastIndexOf(o)
+    if (index >= 0) {
+      inner.remove(index)
+      status += 1
+      true
+    } else
+      false
+  }
+
+  def offer(e: E): Boolean = offerLast(e)
+
+  def remove(): E = removeFirst()
+
+  def poll(): E = pollFirst()
+
+  def element(): E = getFirst()
+
+  def peek(): E = peekFirst()
+
+  def push(e: E): Unit = addFirst(e)
+
+  def pop(): E = removeFirst()
+
+  def size(): Int = inner.size
+
+  private def failFastIterator(startIndex: Int, nex: (Int) => Int) = {
+    new Iterator[E] {
+      private def checkStatus() =
+        if (self.status != actualStatus)
+          throw new ConcurrentModificationException()
+
+      private val actualStatus = self.status
+
+      private var index: Int = startIndex
+
+      def hasNext(): Boolean = {
+        checkStatus()
+        val n = nex(index)
+        (n >= 0) && (n < inner.size)
+      }
+
+      def next(): E = {
+        checkStatus()
+        index = nex(index)
+        inner.get(index)
+      }
+
+      def remove(): Unit = {
+        checkStatus()
+        if (index < 0 || index >= inner.size) {
+          throw new IllegalStateException()
+        } else {
+          inner.remove(index)
+        }
+      }
+    }
+  }
+
+  def iterator(): Iterator[E] =
+    failFastIterator(-1, x => (x + 1))
+
+  def descendingIterator(): Iterator[E] =
+    failFastIterator(inner.size, x => (x - 1))
+
+  override def contains(o: Any): Boolean = inner.contains(o)
+
+  override def remove(o: Any): Boolean = removeFirstOccurrence(o)
+
+  override def clear(): Unit = {
+    if (!inner.isEmpty) status += 1
+    inner.clear()
+  }
+
+  // ArrayList.spliterator is not yet implemented.
+  // def spliterator(): Spliterator[E] = inner.spliterator()
```

Comment:
In general, since spliterators are not implemented anywhere, I recommended not even mentioning it here, even commented. If we one day add spliterators, we'll have to implement them everywhere anyway, in a holistic fashion.

Reply:
Done.